### DISCUSSION TITLE
Error building with gfortran DESPITE using -fallow-argument-mismatch -fallow-invalid-boz

### DISCUSSION THREAD

**Discussion by GrigorisLionis** [esmf-org/esmf#251] [Building / Installing] [answered]
### Requirements

- [X] Reviewed [ESMF User's Guide](https://earthsystemmodeling.org/doc/)
- [X] Searched [GitHub Discussions](https://github.com/orgs/esmf-org/discussions?discussions_q=)

### ESMF Version

v.8.6.1.

### Issue

I am trying to build latest stable version from source (a.k.a. 8.6.1.)
I have encountered a problem compiling the sources, due to a type mismatch error
The output of the error is as following:
gfortran -c -fallow-argument-mismatch -fallow-invalid-boz -O -fPIC  -pthread -ffree-line-length-none  -fopenmp -I/home/user/esmf/esmf-8.6.1/src/Infrastructure/Util/src -I/home/user/esmf/esmf-8.6.1/build_config/Linux.gfortran.default -I/home/user/esmf/esmf-8.6.1/src/Infrastructure -I/home/user/esmf/esmf-8.6.1/src/Superstructure -I/home/user/esmf/esmf-8.6.1/mod/modO/Linux.gfortran.32.mpiuni.default -I/home/user/esmf/esmf-8.6.1/src/include   -DESMF_NO_INTEGER_1_BYTE -DESMF_NO_INTEGER_2_BYTE -DESMF_MPIUNI -I/home/user/esmf/esmf-8.6.1/src/Infrastructure/stubs/mpiuni -DESMF_LOWERCASE_SINGLEUNDERSCORE -DESMF_MOAB=1 -DESMF_LAPACK=1 -DESMF_LAPACK_INTERNAL=1 -DESMF_NO_ACC_SOFTWARE_STACK=1 -DESMF_YAMLCPP=1 -DESMF_YAML=1 -DESMF_NO_OPENACC -DESMF_BOPT_O -DESMF_TESTCOMPTUNNEL -DS32=1 -DESMF_OS_Linux=1 -DESMF_COMM=mpiuni -DESMF_DIR=/home/user/esmf/esmf-8.6.1 /home/user/esmf/esmf-8.6.1/src/Infrastructure/Util/src/ESMF_FortranWordsize.F90 -o /home/user/esmf/esmf-8.6.1/obj/objO/Linux.gfortran.32.mpiuni.default/ESMF_FortranWordsize.o
/home/user/esmf/esmf-8.6.1/src/Infrastructure/Util/src/ESMF_FortranWordsize.F90:425:77:

  425 |  call ESMF_PointerDifference(C_SIZEOF (c_ptrvar), varTK(1), varTK(2), wsize)
      |                                                                             1
Error: Type mismatch in argument ‘n’ at (1); passed INTEGER(4) to INTEGER(8)
/home/user/esmf/esmf-8.6.1/src/Infrastructure/Util/src/ESMF_FortranWordsize.F90:380:77:

  380 |  call ESMF_PointerDifference(C_SIZEOF (c_ptrvar), varTK(1), varTK(2), wsize)
      |                                                                             1
Error: Type mismatch in argument ‘n’ at (1); passed INTEGER(4) to INTEGER(8)
/home/user/esmf/esmf-8.6.1/src/Infrastructure/Util/src/ESMF_FortranWordsize.F90:335:77:

  335 |  call ESMF_PointerDifference(C_SIZEOF (c_ptrvar), varTK(1), varTK(2), wsize)
      |                                                                             1
Error: Type mismatch in argument ‘n’ at (1); passed INTEGER(4) to INTEGER(8)
/home/user/esmf/esmf-8.6.1/src/Infrastructure/Util/src/ESMF_FortranWordsize.F90:290:77:

  290 |  call ESMF_PointerDifference(C_SIZEOF (c_ptrvar), varTK(1), varTK(2), wsize)
      |                                                                             1
Error: Type mismatch in argument ‘n’ at (1); passed INTEGER(4) to INTEGER(8)
/home/user/esmf/esmf-8.6.1/src/Infrastructure/Util/src/ESMF_FortranWordsize.F90:113:79:

  113 |     call ESMF_PointerDifference(C_SIZEOF (c_ptrvar), udt_ptr1, udt_ptr2, psize)
      |                                                                               1
Error: Type mismatch in argument ‘n’ at (1); passed INTEGER(4) to INTEGER(8)
/home/user/esmf/esmf-8.6.1/src/Infrastructure/Util/src/ESMF_FortranWordsize.F90:442:7:

  442 |   use ESMF_FortranWordsizeMod
      |       1
Fatal Error: Cannot open module file ‘esmf_fortranwordsizemod.mod’ for reading at (1): Δεν υπάρχει τέτοιο αρχείο ή κατάλογος
compilation terminated.

I have used a number of different fortran compilers with the same result.

### Autotag

@oehmke

**Comment by danrosen25** [comment]
@GrigorisLionis 
What version of gfortran is being used? Can you confirm that you're setting the build comm layer to 'mpiuni' and the build option to 'O'?

**Reply by GrigorisLionis** [reply]
Hello, thanks for the reply.
![2024-05-29_13-30-50](https://github.com/esmf-org/esmf/assets/135837712/0cd764ec-a2b9-47d8-93a0-36a526698c4d)
Yes, I am building the O version, with ESMF_COMM set to mpiuni.
I am using gfortran-13.3.0, but the same compile error pops up in many different gfortran used 
( 1            /usr/bin/gcc-11/bin/gfortran            80        manual mode
  2            /usr/bin/gfortran-4.4                   50        manual mode
  3            /usr/bin/gfortran-4.6                   60        manual mode
  4            /usr/bin/gfortran-4.9                   40        manual mode
  5            /usr/bin/gfortran-5                     10        manual mode
  6            /usr/bin/gfortran-9                     20        manual mode
  7            /usr/gcc-trunk/bin/gfortran             30        manual mode
  8            /usr/local/gcc-13.3/bin/gfortran-13.3   50        manual mode)

**Comment by danrosen25** [comment]
These compilers have been tested for building ESMF v8.6.1 so I don't think this issue is related to the compiler.
- gfortran 13.2.0
- gfortran 12.2.0
- gfortran 11.2.0
- gfortran 8.3.0
- gfortran 7.4.0

I spoke to the core team and there are two potential system related reasons
- one of the settings provided to the ESMF build system is interfering with the build
- the gfortran built-in spec is interfering with the build by overriding the ESMF build settings

Can you run 'make info' and let me know what it says?

**Comment by GrigorisLionis** [comment]
I am sending you the output of the "make info"
[make.info.txt](https://github.com/esmf-org/esmf/files/15496573/make.info.txt)
The problems  seems to be a mixup of pointer sizes, between c and fortran code, in various parts of the autogenerated code.
The 32bit OS certainly does not help.

**Reply by danrosen25** [reply]
@theurich
Are we testing on any 32-bit ABI systems and do we have access to test on any 32-bit ABI systems?

**Comment by theurich** [comment]
We have not run nightly tests on a 32-bit ABI system in many years. Interacting via this discussion with @GrigorisLionis to get the ESMF internal details for this configuration correct is probably our best bet to move forward here.  Let me take a look at the provided `make.info.txt`...

**Comment by theurich** [comment]
@danrosen25 - I think I see where the 32-bit ABI support in ESMF was broken (back in 2019). This discussion should probably migrate over as an issue to fix the 32-bit ABI support.

**Reply by danrosen25** [reply, marked as answer]
@GrigorisLionis 
If this is the case then it might be some time before we can resolve support for 32 bit ABI. We've opened a ticket to follow up on 32 bit support
https://github.com/esmf-org/esmf/issues/254

**Comment by GrigorisLionis** [comment]
Guys, thank you very much for your answers. I have to say that I stumbled across ESMF looking for a easy way to regrid a NetCDF file  using Python in a legacy machine, and ESMPy looked an excellent choice. It is much easier for my to change to 64 bit... (actually, -layman's opinion- in retrospect it is somewhat peculiar for you  to have , for us to seek, 32 compatibility in this kind of code...)

**Reply by theurich** [reply]
Well, 32-bit ABI support was supposed to work out-of-the-box, so it was a surprise when you reported that it wasn't working. If you could give the branch suggested under https://github.com/esmf-org/esmf/issues/254 a quick test on your 32-bit system, that would help us out greatly. Thanks!

**Reply by danrosen25** [reply]
@GrigorisLionis 
We also have a Docker image with ESMF and ESMPy preinstalled. We've created native images for both arm64 and amd64. I don't know if Docker can emulate amd64 on an i686 machine well but if you have either an arm64 or amd64 machine then you can use the Docker image for small projects.
https://hub.docker.com/r/esmf/esmf-build-release/tags
Once we've resolve the 32 bit ABI issue I can consider building a native image for i686 too.

**Reply by GrigorisLionis** [reply]
Hello, happy to help.

git clone https://github.com/esmf-org/esmf.git --branch v8.6.2b00 --depth 1
I have used this branch


[make.info.txt](https://github.com/user-attachments/files/15531770/make.info.txt)
I am sending you the make info output

Same result as 8.6.1
The compilation stopped on the same file
![Στιγμιότυπο οθόνης_2024-06-03_12-22-30_8 6 2](https://github.com/esmf-org/esmf/assets/135837712/988ded24-291a-4141-9abb-127fd8d66191)

I tried to "hack away"  the easy  problem (INT4->INT8) in that file (a function had a hardcoded 64bit int, and i used a INT8 to cast)
But another problem arose in a subsequent file
![Στιγμιότυπο οθόνης_2024-06-03_12-27-31_8_6_2](https://github.com/esmf-org/esmf/assets/135837712/136209e1-b398-4e4f-8f28-16e76d67798e)
(INT8->INT4)
In that file, the function is defined with an C_INT argument (which presumably is a 32bit in my system) but it is called with an 64bit ESMF_KIND_I8

**Reply by GrigorisLionis** [reply]
CORRECTION
Sorry, have used different 8.6.2

when using
user@PC078 ~/esmf/esmf $ git  clone https://github.com/esmf-org/esmf.git --branch patch/8.6.2
[make.info.txt](https://github.com/user-attachments/files/15534662/make.info.txt)


the compilation is stuck at the second file 
![Στιγμιότυπο οθόνης_2024-06-03_16-41-33_8 6 2 patch](https://github.com/esmf-org/esmf/assets/135837712/fcc21425-2ecc-4774-b099-6a142efa8855)

**Reply by theurich** [reply]
@GrigorisLionis - yes `--branch patch/8.6.2` is what you want! Good to know about the next issue in `ESMF_Info` when you test this on your 32-bit system. I will take a look at that one next and let you know.
I suggest that we move to https://github.com/esmf-org/esmf/issues/254 as we iterate on the details of this fix.
